moveit_ros/trajectory_cache/src/test_trajectory_cache.cpp

```diff
@@ -0,0 +1,1066 @@
+      waypoints.at(i).position.y = i;
+      waypoints.at(i).position.z = i;
+      waypoints.at(i).orientation.w = i;
+      waypoints.at(i).orientation.x = i + 0.1;
+      waypoints.at(i).orientation.y = i + 0.1;
+      waypoints.at(i).orientation.z = i + 0.1;
+    }
+    return waypoints;
+  }();
+  return out;
+}
+
+void test_motion_trajectories(std::shared_ptr<MoveGroupInterface> move_group, std::shared_ptr<TrajectoryCache> cache)
+{
+  // Setup =====================================================================
+  // All variants are copies.
+
+  /// MotionPlanRequest
+
+  // Plain start
+  moveit_msgs::msg::MotionPlanRequest plan_req;
+  move_group->constructMotionPlanRequest(plan_req);
+  plan_req.workspace_parameters.header.frame_id = g_robot_frame;
+  plan_req.workspace_parameters.max_corner.x = 10;
+  plan_req.workspace_parameters.max_corner.y = 10;
+  plan_req.workspace_parameters.max_corner.z = 10;
+  plan_req.workspace_parameters.min_corner.x = -10;
+  plan_req.workspace_parameters.min_corner.y = -10;
+  plan_req.workspace_parameters.min_corner.z = -10;
+  plan_req.start_state.multi_dof_joint_state.joint_names.clear();
+  plan_req.start_state.multi_dof_joint_state.transforms.clear();
+  plan_req.start_state.multi_dof_joint_state.twist.clear();
+  plan_req.start_state.multi_dof_joint_state.wrench.clear();
+
+  // Empty frame start
+  moveit_msgs::msg::MotionPlanRequest empty_frame_plan_req = plan_req;
+  empty_frame_plan_req.workspace_parameters.header.frame_id = "";
+
+  // is_diff = true
+  auto is_diff_plan_req = plan_req;
+  is_diff_plan_req.start_state.is_diff = true;
+  is_diff_plan_req.start_state.joint_state.header.frame_id = "";
+  is_diff_plan_req.start_state.joint_state.name.clear();
+  is_diff_plan_req.start_state.joint_state.position.clear();
+  is_diff_plan_req.start_state.joint_state.velocity.clear();
+  is_diff_plan_req.start_state.joint_state.effort.clear();
+
+  // Something close enough (mod 0.1 away)
+  auto close_matching_plan_req = plan_req;
+  close_matching_plan_req.workspace_parameters.max_corner.x += 0.05;
+  close_matching_plan_req.workspace_parameters.min_corner.x -= 0.05;
+  close_matching_plan_req.start_state.joint_state.position.at(0) -= 0.05;
+  close_matching_plan_req.start_state.joint_state.position.at(1) += 0.05;
+  close_matching_plan_req.start_state.joint_state.position.at(2) -= 0.05;
+  close_matching_plan_req.goal_constraints.at(0).joint_constraints.at(0).position -= 0.05;
+  close_matching_plan_req.goal_constraints.at(0).joint_constraints.at(1).position += 0.05;
+  close_matching_plan_req.goal_constraints.at(0).joint_constraints.at(2).position -= 0.05;
+
+  // Close with swapped constraints (mod 0.1 away)
+  auto swapped_close_matching_plan_req = close_matching_plan_req;
+  std::swap(swapped_close_matching_plan_req.goal_constraints.at(0).joint_constraints.at(0),
+            swapped_close_matching_plan_req.goal_constraints.at(0).joint_constraints.at(1));
+
+  // Smaller workspace start
```

Comment:
A cache hit is valid if the requested planning constraints has a workspace that completely subsumes a cached plan's workspace.

E.g. , if workspace is defined by the extrema (x_min, y_min, x_max, y_max)

Then if:
Request: (-1, -1, 1, 1)
Plan in cache: (-0.5, -0.5, 0.5, 0.5) (potential valid match)

But
Plan in cache: (-2, -0.5, 0.5, 0.5) (not valid, since this plan might cause the end effector to go out of bounds.